lib/puppet_x/puppetlabs/powershell/powershell_manager.rb

```diff
@@ -64,9 +64,14 @@ def initialize(cmd, debug)
             sleep 0.2
           end
         end
-
-        fail "Failure waiting for PowerShell process #{@ps_process[:pid]} to start pipe server" if @pipe.nil?
-
+        if @pipe.nil?
+          # Tear down and kill the process if unable to connect to the pipe; failure to do so
+          # results in zombie processes being left after the puppet run.
+          @stdout.close if !@stdout.closed?
+          @stderr.close if !@stderr.closed?
+          Process.kill("KILL", @ps_process[:pid])
```

Comment:
So one question that arises is... should we grab the process handle immediately after starting the process, as the system may reuse PIDs (assuming `@ps_process` doesn't already have a handle)? If the process fails to start in the first place or quickly crashes and burns, we've now waited 30 seconds, which increases the likelihood that the PID has been reassigned.  That said, PIDs cannot be reused until all handles to the process are closed - and I'm not sure what state Ruby is actually holding on to at this point - so we may have to look to the Ruby source for clarity.

We should be able to use an API call to `OpenProcess` with `PROCESS_ALL_ACCESS`, `TRUE`, `@ps_process[:pid]` to get a handle, which we can then use to kill the process here... but that may be overkill (pardon the pun)

Reply:
`popen3` returns the same return type as [Process.detach](https://ruby-doc.org/core-2.1.1/Process.html#method-c-detach) at https://github.com/ruby/ruby/blob/trunk/lib/open3.rb#L200

The nominal implementation of `Process.detach` lives in the Ruby source at https://github.com/ruby/ruby/blob/fe3decb2017af2092d4b2181e27e4cd8f6bbf4fa/process.c#L1086-L1093

The Ruby code gets a little squirrely though as the implementations here use `waitpid` and `wait4`, but I believe the Windows variant is backed by the code at https://github.com/ruby/ruby/blob/e33b1690d06f867e45750bd8e3e8b06d78b5bc26/win32/win32.c#L4496-L4569. You'll also find the expected calls to `GetExitCodeProcess` here.... and each of those calls seems to be relying on `child->hProcess`.  That makes me think Ruby is holding on to a handle, even if it isn't exposed in its APIs directly anywhere.